### DISCUSSION TITLE
Chapter 4: Functional programming in Scala 3

### DISCUSSION THREAD

**Discussion by gvolpe** [gvolpe/trading#15] [General]
## Chapter 4: Functional programming in Scala 3

⚠️ Note that the order and number of chapters is not yet fully decided, so expect changes here as well ⚠️

**Comment by kyri-petrou** [comment]
Hey @gvolpe, really enjoying the book so far!

I have a question regarding the `Newtype` definitions (apologies in advance if it's mentioned somewhere and I missed it!)

Is there a specific reason that you're using an `abstract class` instead of a `trait`?  I'm not too well across the internals of Scala so I'm not sure whether there might be some overhead involved with traits that's not an issue when using abstract classes

**Comment by gvolpe** [comment]
Hi @kyri-petrou ,

When I first wrote that code, there was a slight difference in the generated byte code that made one box and the other not, IIRC. I think that has changed in recent versions of Scala 3, so probably using `trait` gives us the same results.

Either way, that's code that should belong in a library, given the lack of support for newtypes in Scala 3. I hope we get something equivalent to `scala-newtype` at some point, so this code can disappear :)

**Reply by kyri-petrou** [reply]
Okay cool. Yeah I also had a look at the generated byte code and didn't see any difference (albeit I'm not entirely sure how to identify when something becomes boxed).

So assuming that it's safe to use traits, could we mix them in when creating new instances of `Newtype` to add functionality? e.g., something like:

``` scala
trait Newtype[A]:
  opaque type Type = A
  inline def apply(a: A): Type                                     = a
  extension (t: Type) inline def value: A                          = t
  protected inline final def derive[F[_]](using ev: F[A]): F[Type] = ev
end Newtype

trait CirceSerialization[A](using enc: Encoder[A], dec: Decoder[A]):
  opaque type Type = A
  given Encoder[Type] = enc
  given Decoder[Type] = dec
end CirceSerialization
```

and then use it with:

``` scala
object PlainString extends Newtype[String]
type PlainString = PlainString.Type

object JsonString extends Newtype[String] with CirceSerialization[String]
type JsonString = JsonString.Type
```

Of course in the example above we could have just extended `Newtype` in the definition, but I was referring to a case where we have many different traits that we can extend our Newtype with and we want to mix multiple in depending on how the opaque type will be used.

And yeah I know, it would be awesome if we could get a Scala 3 version of `newtype-scala` or similar so we don't have to write this code ourselves. Fingers crossed it won't be too long 🤞 

PS: We're starting a new project at work using Scala 3 which will pretty much utilize 80% of everything you covered in the book, so it couldn't have been better timed! Thanks for the awesome work!

**Reply by gvolpe** [reply]
That sounds super cool! Have you considered [Monix Newtypes](https://github.com/monix/newtypes) as well (also mentioned in the book)? It's not for everyone, and a custom encoding allows you to put default instances you want derived for all newtypes, but worth having a look into it.

**Reply by kyri-petrou** [reply]
Ah I saw that sentence but for some reason I thought it was an old library, didn't realise it was for Scala 3. Thanks for that, I'll look into it! It seems that it might be a good fit!

On a side-note, what's your preferred method for staying up to speed with recent Scala 3 developments? I've been "Watching" repos of interest on github but I was wondering if there's a board or channel anywhere that people put announcements on new libraries / releases / etc

**Reply by gvolpe** [reply]
Yes, it works for both Scala 2 and 3, so it's great if you have shared code with different versions!

Nothing I can recommend, unfortunately. It's been a lot of manual work so far, following the right people on Github, Twitter, etc, looking at Scala newsletters, and so on.

**Reply by kyri-petrou** [reply]
Yeah that does sound like quite a bit of work.

Thanks heaps again for the help!

**Reply by cihleman3** [reply]
Would `object JsonString extends Newtype[String] with CirceSerialization[String]` actually work? I thought opaque types cannot be overwritten.
